### ISSUE TITLE
Make use of qiskit.test.providers.ProviderTestCase

### ISSUE THREAD

**Issue by diego-plan9** [Qiskit/qiskit-ibmq-provider#52]
### What is the expected behavior?

Since a number of commits ago, terra includes [some classes for helping testing providers and backends](https://github.com/Qiskit/qiskit-terra/tree/master/qiskit/test/providers). Ideally this provider should make use of those for consistency - it involves changing the class the relevant test cases inherit from, and removing individual tests as necessary.

**Comment by jwoehr** [comment]
So the new class would be `class IBMQProviderTestCase(ProviderTestCase)` ?

**Comment by diego-plan9** [comment]
Exactly - the idea is to make use of inheriting from `BackendTestCase` and `ProviderTestCase`, and in the process revise the tests of each test case individually, removing those that are already covered (as in, they test "the same thing") by those base classes (as they include a minimal set of tests) to avoid duplication.

**Comment by jwoehr** [comment]
I'm working on this now.

**Comment by diego-plan9** [comment]
Yes, the decorators can and should be used.

**Comment by jwoehr** [comment]
`requires_qe_access` is not a big problem. But qiskit/test/providers/backend.py seems oriented to a single backend, rather than a list of backends. `qiskit-ibmq-provider/test/ibmq/test_ibmq_backends.py` is designed around such a list. Perhaps I should factor out the testing of an individual backend from `/test_ibmq_backends.py` and not make that file itself inherit `IBMQBackendTestCase`.

**Comment by jwoehr** [comment]
Perhaps  `qiskit-ibmq-provider/test/ibmqtest_ibmq_backends.py` should enumerate the QE backends and dynamically create a  `class IBMQBackend_backendname_TestCase(IBMQBackendTestCase)` for each enumerated instance and `run()` it ... `IBMQBackendTestCase` I have already defined as
```
from qiskit.test.providers import BackendTestCase
class IBMQBackendTestCase(BackendTestCase):
```
... etc.

**Comment by jwoehr** [comment]
I understand the structure of the test cases better now and am proceeding with coding.

**Comment by diego-plan9** [comment]
Great, looking forward to the results! Feel free to handle both in different PRs (ie. one for the provider test, and another one for the backend tests, which are probably the most complex as you mention) if it makes it more manageable - specially considering we will be having a number of changes in the upcoming couple of weeks and might involve a small burst of activity.

**Comment by jwoehr** [comment]
This issue requires refactoring the test cases themselves. The BackendTestClass class is biased towards backend classes that can be instantiated rather simply. The actual test coverage for ibmq is easiest retrieving the Backend instances returned by IBMQ.backends() and testing them on the fly. I'm working on that.

Apologies if it's going a little slow right now, I've gotten busy, but am working on it some every day.

**Comment by jwoehr** [comment]
Is the "burst of activity" you mention @diego-plan9 going to involve modifying the test cases in any way?

**Comment by diego-plan9** [comment]
No rush at all @jwoehr  - and thanks for working in the issue, we are glad to receive the contribution whenever it suits your schedule!

> Is the "burst of activity" you mention @diego-plan9 going to involve modifying the test cases in any way?

It will likely add some new tests to existing files or small tweaks here and there, hopefully (the most likely candidate of having a slightly bigger number of changes would be `test_ibmq_connector.py`). The suggestion of splitting the PR into smaller ones was minimizing the overhead of conflicts in files - but even if they occur, I'm confident we will be able to sort them out, so don't worry too much about them!

**Comment by jwoehr** [comment]
w/r/t splitting the pr, it seems to me that the current factoring of test cases overlaps the line drawn by `ProviderTestCase` and `BackendTestCase` so that the implementation of either rebase implies the implementation of both. Thus I am working on sorting it all out. I was wondering if people are currently modifying the extant test cases in `qiskit-ibmq-provider/test/ibmq/` ... your answer seems to be "yes", so maybe I should chat with whoever is doing that work.

**Comment by jwoehr** [comment]
Is there, in fact, a deadline, a target data for next release?

**Comment by jwoehr** [comment]
I've figured it out and have a new `class IBMQBackendTestCase(BackendTestCase)` working.

Now I have to change `class TestIBMQBackends` to inherit from `class IBMQProviderTestCase(ProviderTestCase)`

**Comment by diego-plan9** [comment]
> Is there, in fact, a deadline, a target date for next release?

There will probably be a small one this week to sync with the terra release, and the following one still to be decided (likely ~2 weeks) - but this issue should be relatively independent of the releases, just feel free to ignore them!

**Comment by diego-plan9** [comment]
After the latest conversations at #68, updated the issue to split in two. This one will be used for keeping track of making use `ProviderTestCase`, and a new one will be used for the trickier `BackendTestCase`.

**Comment by jwoehr** [comment]
Awesome, will address after I return from COMMON conference.

**Comment by diego-plan9** [comment]
Closed via #124
